src/traffic_server/InkAPI.cc

```diff
@@ -3262,43 +3262,24 @@ TSMimeHdrFieldCopyValues(TSMBuffer dest_bufp, TSMLoc dest_hdr, TSMLoc dest_field
   return TS_SUCCESS;
 }
 
-// TODO: This is implemented horribly slowly, but who's using it anyway?
-//       If we threaded all the MIMEFields, this function could be easier,
-//       but we'd have to print dups in order and we'd need a flag saying
-//       end of dup list or dup follows.
 TSMLoc
 TSMimeHdrFieldNext(TSMBuffer bufp, TSMLoc hdr, TSMLoc field)
 {
   sdk_assert(sdk_sanity_check_mbuffer(bufp) == TS_SUCCESS);
   sdk_assert((sdk_sanity_check_mime_hdr_handle(hdr) == TS_SUCCESS) || (sdk_sanity_check_http_hdr_handle(hdr) == TS_SUCCESS));
   sdk_assert(sdk_sanity_check_field_handle(field, hdr) == TS_SUCCESS);
 
-  MIMEFieldSDKHandle *handle = (MIMEFieldSDKHandle *)field;
-
-  if (handle->mh == nullptr) {
-    return TS_NULL_MLOC;
-  }
-
-  int slotnum = mime_hdr_field_slotnum(handle->mh, handle->field_ptr);
-  if (slotnum == -1) {
-    return TS_NULL_MLOC;
-  }
-
-  while (true) {
-    ++slotnum;
-    MIMEField *f = mime_hdr_field_get_slotnum(handle->mh, slotnum);
```

Comment:
Returning iterator looks great! 

But I'm wondering about performance improvement. It looks like `MIMEHdrImpl::iterator::operator++()` goes `step()` and does almost the same search. Where is the shortcut?

https://github.com/apache/trafficserver/blob/664e80eaf392f36ece50cfc463bb22dc25117395/proxy/hdrs/MIME.cc#L1285-L1312

https://github.com/apache/trafficserver/blob/664e80eaf392f36ece50cfc463bb22dc25117395/proxy/hdrs/MIME.h#L351-L364

Reply:
It starts in the correct `block`. Note there is no reference to `m_first_fblock`. The old style always searches from there. If the iterator is currently not on the last field in the current block and the next field is valid, the loop only has one execution, i.e. it is constant time independent of the number of blocks in the header. Even at the end of a block the only extra cost is a single block advance and again, it's constant time. The only case where the iteration is not constant time is if there are dead fields. However, if we switched to a free list style as noted in #8466 even that would be minimized.

Reply:
It starts in the correct `block`. Note there is no reference to `m_first_fblock`. The old style always searches from there. If the iterator is currently not on the last field in the current block and the next field is valid, the loop only has one execution, i.e. it is constant time independent of the number of blocks in the header. Even at the end of a block the only extra cost is a single block advance and again, it's constant time. The only case where the iteration is not constant time is if there are dead fields. However, if we switched to a free list style as noted in #8466 even that would be minimized.

Reply:
Further, in the old case, if there is a dead field, _it starts the search over again at the first block even though it has found the block with the current field in it_. This makes iterating over headers with multiple dead fields seriously non-performant.

Reply:
Cool, this search will be removed. Thanks for the clarification. 
```
     fblock = &(mh->m_first_fblock); 
     while (block_num-- && fblock) { 
       fblock = fblock->m_next; 
     } 
```

Let's figure out what's the best approach to avoid the heap allocation completely from these TS APIs. But that's another story for another time